rfcs/taskforce/0001-buildkiteci.md

```diff
@@ -0,0 +1,243 @@
+
+P4 → BuildKite agents can be installed and tagged locally (or on your own instances) to iterate on builds on your own hardware before putting it into production
+
+P5 → Flakiness is explicit in the Dhall configuration
+
+P6 → Nix builds work inside docker containers "on my machine" so there is no evidence that they won't work on gcloud.
+
+## Mapping of resource utilization to tasks
+
+Big machines — 8core, 30GB of RAM
+
+- Builds
+- Heavy integration tests
+- All unit tests
+
+Light machines — 1core, 3.75GB of RAM
+
+- Linting
+- Small tools
+- (when we break up unit tests)
+
+## Projected Spend
+
+We can run 20 Big machines for ~80% of what we're paying circle per month even if these 20 machines were on 100% of the time and not preemptible.
+
+This will also become much cheaper when we do autoscaling or if we use preemptible instances and support retries in case of preempt failure.
+
+If necessary we can look into getting this system to work on Azure to make better use of our credits, but Buildkite works much better with Gcloud and/or Aws. Azure credits will be easily used up by compute nodes during QA-net debugging.
+
+## Rough plan of moving over
+
+Work can be parallelized across working on the deploy infrastructure, pipeline creation, artifact management work
+
+1. Start with a super easy job (lint)
+2. Enable on both Circle and Buildkite and disable on circle only after 3 working days of PRs
+3. Move over a very high impact job — build-artifacts-medium-curves + all dependencies. Parallelize the crap out of it and set us up to reuse artifacts for other jobs.
+4. Put in the new integration tests (blocked on implementing the new integration tests)
+
+=== This is the point at which we can stop for task force (or maybe we'll want 5) ===
+
+5. Split unit tests into more granular parallel jobs
+
+6. Use pre-emptible instances
+
+7. Auto-scaling
+
+8. Move the rest of the work over other than mac builds
+
+9. Handling mac builds ourselves either with an in-house mac machine or with mac in a data center
+
+## Support for external contributors
+
+With build machines running on our infrastructure we need to be careful about
+running arbitrary code from others. Thanks @yourbuddyconner for flagging this
+one.
+
+In order to keep our systems safe, CI should not start immediately for non-core
+team members but rather only start when a core team kicks off the build. This
+can be triggered initially by a step in buildkite, and later move to a bot in
+GitHub.
+
+The nice thing about this approach is we no longer need to support a
+"second class" version of the build for non-core team members that disables
+access to necessary secrets. We can assume all builds have access to all
+variables.
+
+Core team reviewers will of course need to double check to see if any secrets
+are trying to be exfiltrated or our infrastructure is attempting to be
+compromised before granting permission for CI to run.
```

Comment:
We need to be *very* careful about how we implement this. If we approve a pull request for CI, we need to be sure that subsequent commits **do not** also get run automatically, or the contributors can push more commits to their branch to exfiltrate the secrets.

Reply:
Some logic similar to Rust's `bors` PR approval logic would likely be useful: https://bors.tech/documentation/getting-started/ (ctrl+f `Adding reviewers`)